tests/testthat/test-function_return_linter.R

```diff
@@ -0,0 +1,68 @@
+test_that("function_return_linter skips allowed usages", {
+  lines_simple <- trim_some("
+    foo <- function(x) {
+      x <- x + 1
+      return(x)
+    }
+  ")
+  expect_lint(lines_simple, NULL, function_return_linter())
+
+  # arguably an expression as complicated as this should also be assigned,
+  #   but for now that's out of the scope of this linter
+  lines_subassignment <- trim_some("
+    foo <- function(x) {
+      return(x[, {
+        col <- col + 1
+        .(grp, col)
+      }])
+    }
+  ")
+  expect_lint(lines_subassignment, NULL, function_return_linter())
+})
+
+test_that("function_return_linter blocks simple disallowed usages", {
+  linter <- function_return_linter()
+  lint_msg <- rex::rex("Move the assignment outside of the return() clause")
+
+  expect_lint(
+    trim_some("
+      foo <- function(x) {
+        return(x <- x + 1)
+      }
+    "),
+    lint_msg,
+    linter
+  )
+
+  expect_lint(
+    trim_some("
+      foo <- function(x) {
+        return(x <<- x + 1)
+      }
+    "),
+    lint_msg,
+    linter
+  )
```

Comment:
Also add this one for the sake of completeness?

``` r
library(lintr)

lint(
  text = "foo <- function(x) {
        return(x + 1 ->> x)
      }
    ",
  linters = function_return_linter()
)
#> <text>:2:16: warning: [function_return_linter] Move the assignment outside of the return() clause, or skip assignment altogether.
#>         return(x + 1 ->> x)
#>                ^~~~~~~~~~~
```

<sup>Created on 2022-09-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>